### ISSUE TITLE
dokku_client.sh - Allow for selection of APP based on DOKKU_HOST/DOKKU_GIT_REMOTE

### ISSUE THREAD

**Issue by kbuilds** [dokku/dokku#5564]
### Description of feature

In older versions of Dokku, I was able to have several different git remotes for the same repository which each pointed to different Dokku apps. These different apps could be running on different Dokku instances, or on the same instance. The Dokku client would figure out which host and app I was using based on the different git remotes, and based on the DOKKU_HOST environment variable.

For example I would have the following git remotes:

stage: dokku@stage.myapp.com:my-app-stage
prod: dokku@www.myapp.com:my-app-prod

If I wanted my Dokku client to talk to the staging app, I would run:

```
export DOKKU_HOST='stage.myapp.com'
dokku_client.sh config:show
```

If I wanted  my Dokku client to talk to the production app, I would run:

```
export DOKKU_HOST='www.myapp.com'
dokku_client.sh config:show
```

The app name was being parsed from whichever git remote contained the value of the `DOKKU_HOST` environment variable. However, it looks like this behavior was changed in this commit:

https://github.com/dokku/dokku/commit/b7c71558522d840056ce4455ce34d21c83e640eb

I see that this change was made in response to the `--remote` flag not being respected when the same hostname is used for multiple remotes:

https://github.com/dokku/dokku/issues/4838

The old behavior is useful because it allows the `DOKKU_HOST` environment variable to be set once, then all subsequent Dokku client commands are directed at that host/app.

I think that it might be useful to have the client check for the presence of the `--remote` flag, and if it's not found, then use the value of a `DOKKU_GIT_REMOTE` environment variable. That way, users of the client can have the original behavior by setting that variable.

I am not a bash expert, but I think that this could be accomplished simply by changing this line:

https://github.com/dokku/dokku/blob/master/contrib/dokku_client.sh#L66

```
main() {
  declare CMD="$1" APP_ARG="$2"
  local APP="" DOKKU_GIT_REMOTE="dokku" DOKKU_REMOTE_HOST=""
  local cmd_set=false next_index=1 skip=false args=("$@")

  for arg in "$@"; do
```

To this

```
main() {
  declare CMD="$1" APP_ARG="$2"
  local APP="" DOKKU_GIT_REMOTE=${DOKKU_GIT_REMOTE:="dokku"} DOKKU_REMOTE_HOST=""
  local cmd_set=false next_index=1 skip=false args=("$@")

  for arg in "$@"; do
```

However, I am not sure if there might be other consequences of making this change. If this isn't going to totally break the client, I would be happy to make a PR.

**Comment by josegonzalez** [comment]
Why is setting the --remote flag not an option?

**Comment by josegonzalez** [comment]
Also, how does heroku handle this sort of issue with their cli client?

**Comment by josegonzalez** [comment]
Seems like it uses our --remote method https://devcenter.heroku.com/articles/multiple-environments

**Comment by josegonzalez** [comment]
Maybe we could use a git config for that? https://github.com/heroku/legacy-cli/issues/169#issuecomment-7466829

So basically call `dokku remote:set staging` which calls the respective `git config` command. Then when we call dokku commands, check that config and use that remote, otherwise fallback to the default dokku remote?

We could also have a `remote:unset` call that resets the value?

**Comment by kbuilds** [comment]
@josegonzalez 

>Why is setting the --remote flag not an option?

It is an option, but when sending multiple sequential commands to the same environment, it just becomes a little tedious. I could also wrap these commands into a local script which adds that `--remote` flag.

>Also, how does heroku handle this sort of issue with their cli client?

I took a look at the two links you provided, and I will take a closer look into how Heroku handles this, but it looks like Heroku relies on a combination of:

- Setting a `HEROKU_APP` env var
- Setting a `remote` directive in the repo's git config
- Using the `--remote` flag

I wouldn't expect the Dokku client to function any differently from the Heroku client, except where absolutely necessary. Worst case scenario, I will just create a wrapper around dokku_client.sh to handle settings. But let me mess around with the Heroku client before I close this.

**Comment by josegonzalez** [comment]
I think setting DOKKU_GIT_REMOTE would be effectively the same as HEROKU_APP, so if your change makes that usable, that is fine by me.

**Comment by kbuilds** [comment]
@josegonzalez I don't think that they are quite the same, since `DOKKU_GIT_REMOTE` specifies the name of a local git remote, while `HEROKU_APP` specifies the name of a Heroku app located at whichever remote matches. **I think**

I am going to do a little more research on this. I am having trouble finding references to the `HEROKU_APP` environment variable in their docs.

**Comment by kbuilds** [comment]
@josegonzalez I did a little more research on this, and it looks like my assumption in my previous comment is correct. Upon inspecting the source code in the Heroku CLI, I was able to see how the `HEROKU_APP` is actually being used, and it is basically the same as setting the `--app` flag on each command:

Reference: https://github.com/heroku/cli/search?q=HEROKU_APP

In each source file, the value for the app name is derived from the `--app` CLI flag, and if that is not found, then it looks for the value of the `HEROKU_APP` env var.

Back to our problem:

Since there is no Heroko equivalent of `DOKKU_GIT_REMOTE`, I don't think that it makes sense to adopt the change that I proposed above.

If I was facing the same issue with Heroku, this was probably the most interesting solution that I found:

https://devcenter.heroku.com/articles/multiple-environments#specifying-an-environment-with-the-heroku-cli

This means that the Heroku CLI will send commands to whichever remote is specified for "heroku" in the repo's git config will be used for all commands.

This is what a sample git config could look like when using Dokku:

.git/config:

```
[remote "prod"]
        url = dokku@www.myapp.com:my-app-prod
        fetch = +refs/heads/*:refs/remotes/prod/*
[remote "stage"]
        url = dokku@stage.myapp.com:my-app-stage
        fetch = +refs/heads/*:refs/remotes/stage/*
[dokku]
        remote = prod
```

The above config, would effectively make every command the same as running:

```
dokku_client.sh --remote prod <command>
```

without the need to specify the `--remote` flag on each command.

This would definitely work for my use-case. Unfortunately, I wouldn't even know where to start implementing this in bash, so I would be useless for a PR.
